lib/charms/opensearch/v0/helper_cluster.py

```diff
@@ -77,106 +75,39 @@ def get_cluster_settings(
     @staticmethod
     def recompute_nodes_conf(app_name: str, nodes: List[Node]) -> Dict[str, Node]:
         """Recompute the configuration of all the nodes (cluster set to auto-generate roles)."""
-        # in case the cluster nodes' roles were previously "manually generated" - we need
-        # to reset the roles to their expected default values so that "roles re-balancing"
-        # logic (node_with_new_roles) can be safely applied: only change the last node.
-        # Nothing changes to the conf of the nodes if the roles were previously auto-generated.
-        valid_nodes = ClusterTopology.refill_node_with_default_roles(app_name, nodes)
-
-        nodes_by_name = dict([(node.name, node) for node in nodes])
-
-        # compute node with new roles - only returns 1 changed node
-        updated_node = ClusterTopology.node_with_new_roles(app_name, valid_nodes)
-        if updated_node:
-            nodes_by_name[updated_node.name] = updated_node
-
-        return nodes_by_name
-
-    @staticmethod
-    def refill_node_with_default_roles(app_name: str, nodes: List[Node]) -> List[Node]:
-        """Refill the roles of a list of nodes with default values for re-computing.
-
-        This method works hand in hand with node_with_new_roles which assumes a clean
-        base (regarding the auto-generation logic) and only applies changes to 1 node.
-        """
+        if not nodes:
+            return {}
+        logger.debug(f"Roles before re-balancing {({node.name: node.roles for node in nodes})=}")
+        nodes_by_name = {}
+        current_cluster_nodes = []
+        for node in nodes:
+            if node.app_name == app_name:
+                current_cluster_nodes.append(node)
+            else:
+                # Leave node unchanged
+                nodes_by_name[node.name] = node
         base_roles = ["data", "ingest", "ml", "coordinating_only"]
         full_roles = base_roles + ["cluster_manager"]
-
-        current_cluster_nodes = [node for node in nodes if node.app_name == app_name]
-        current_cm_eligible = [node for node in current_cluster_nodes if node.is_cm_eligible()]
-
-        # we check if the roles were previously balanced in accordance with the auto-generation
-        # logic, in which the max difference between CMs and Non-CMs is 1 node (to keep quorum)
-        unbalanced = len(current_cm_eligible) < len(current_cluster_nodes) - 1
-
-        updated_nodes = []
+        highest_unit_number = max(node.unit_number for node in current_cluster_nodes)
         for node in current_cluster_nodes:
             # we do this in order to remove any non-default role / add any missing default role
-            new_roles = full_roles if unbalanced or node.is_cm_eligible() else base_roles
-            updated = Node(
+            if len(current_cluster_nodes) % 2 == 0 and node.unit_number == highest_unit_number:
```

Comment:
@carlcsaposs-canonical nice! Really good simplification! 

I see 2x potential issues we need to address:
1) What happens if I deploy 15 nodes at once? I will have a first run with `len(current_cluster_nodes) == 1`, then `len(...)==2`, etc. Also, this logic is called [here](https://github.com/canonical/opensearch-operator/blob/c0225203320a412c3492ac827ac47b3b5b763787/lib/charms/opensearch/v0/opensearch_base_charm.py#L341) and it will be re-executed multiple times while the cluster is deploying. Why not using `planned_units` here instead? That will return `15` every time. Btw, this is going to be an issue at both start-up and any time we do a scale up.
2) If I have a cluster with 2 nodes and let's assume `opensearch/1` is the charm leader; then I add 3 more nodes. Now, `opensearch/1` will get a new role ("cluster_manager"). However, if [you look in this code path](https://github.com/canonical/opensearch-operator/blob/c0225203320a412c3492ac827ac47b3b5b763787/lib/charms/opensearch/v0/opensearch_base_charm.py#L344), the leader never requests a restart. I think we need to update this line from `elif app_data`, to `if app_data`.

I guess we will not see (2) at deployment time, as we keep deferring the `start` event, which at its end, has a restart request. However, as we need to expand the cluster, then I think (2) may become an issue.

Reply:
@carlcsaposs-canonical makes sense. The `planned_units` is used along the path, so you will get the total number of units anyway.